File: app/src/main/java/com/adroitandroid/p2pchat/ChatAdapter.java
package com.adroitandroid.p2pchat;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.adroitandroid.p2pchat.databinding.RowMessagesBinding;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by pv on 22/06/17.
 */

class ChatAdapter extends RecyclerView.Adapter<ChatAdapter.MessageVH> {
    private final List<String> mMessages;
    private final List<String> mMessageSender;

    ChatAdapter() {
        mMessages = new ArrayList<>();
        mMessageSender = new ArrayList<>();
    }

    @Override
    public MessageVH onCreateViewHolder(ViewGroup parent, int viewType) {
        RowMessagesBinding binding = RowMessagesBinding.inflate(
                LayoutInflater.from(parent.getContext()), parent, false);
        return new MessageVH(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull MessageVH holder, int position) {
        String sender = mMessageSender.get(position);
        holder.binding.setSender(sender);
        holder.binding.setMessage(mMessages.get(position));
        holder.binding.setUserIsSender("you".equalsIgnoreCase(sender));
        holder.binding.executePendingBindings();
    }

    @Override
    public int getItemCount() {
        return mMessages.size();
    }

    void addMessage(String message, String senderName) {
        mMessages.add(message);
        mMessageSender.add(senderName);
        notifyItemInserted(mMessages.size() - 1);
    }

    class MessageVH extends RecyclerView.ViewHolder {

        private final RowMessagesBinding binding;

        MessageVH(@NonNull RowMessagesBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }
}

File: app/src/main/java/com/adroitandroid/p2pchat/ChatAdapter.kt
package com.adroitandroid.p2pchat

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.adroitandroid.p2pchat.databinding.RowMessagesBinding
import java.util.*

/**
 * Created by pv on 22/06/17.
 */
internal class ChatAdapter : RecyclerView.Adapter<ChatAdapter.ViewHolder>() {
    private val messages: MutableList<String> = ArrayList()
    private val messageSenders: MutableList<String> = ArrayList()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ChatAdapter.ViewHolder {
        val binding = RowMessagesBinding.inflate(
                LayoutInflater.from(parent.context), parent, false)
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) = with(holder.binding) {
        val messageSender = messageSenders[position]
        sender = messageSender
        message = messageSenders[position]
        userIsSender = "you".equals(sender, ignoreCase = true)
        executePendingBindings()
    }

    override fun getItemCount(): Int {
        return messageSenders.size
    }

    fun addMessage(message: String, senderName: String) {
        messageSenders.add(message)
        messageSenders.add(senderName)
        notifyItemInserted(messages.size - 1)
    }

    internal inner class ViewHolder(val binding: RowMessagesBinding) : RecyclerView.ViewHolder(binding.root)
}